src/main/java/ch/njol/skript/expressions/ExprBanData.java

```diff
@@ -0,0 +1,226 @@
+			banEntry = getBanEntry(playerTarget);
+		else
+			return null;
+
+		if (banEntry == null)
+			return null;
+
+		return switch (entryType) {
+			case BAN_DATE -> {
+				java.util.Date creation = banEntry.getCreated();
+				ch.njol.skript.util.Date skriptCreation = new ch.njol.skript.util.Date(creation.getTime());
+				yield new Object[]{ skriptCreation };
+			}
+			case SOURCE -> new Object[]{ banEntry.getSource() };
+			case EXPIRE_DATE -> {
+				java.util.Date expiration = banEntry.getExpiration();
+				if (expiration == null)
+					yield null;
+				ch.njol.skript.util.Date skriptExpiration = new ch.njol.skript.util.Date(expiration.getTime());
+				yield new Object[]{ skriptExpiration };
+			}
+			case REASON -> new Object[]{ banEntry.getReason() };
+		};
+	}
+
+	@Override
+	public boolean isSingle() {
+		return true;
+	}
+
+	@Override
+	public Class<?> getReturnType() {
+		return switch (entryType) {
+			case BAN_DATE, EXPIRE_DATE -> ch.njol.skript.util.Date.class;
+			case SOURCE, REASON -> String.class;
+		};
+	}
+
+	@Override
+	public Class<?> @Nullable [] acceptChange(ChangeMode mode) {
+		if (mode == ChangeMode.SET) {
+			return switch (entryType) {
+				case EXPIRE_DATE -> new Class<?>[]{ ch.njol.skript.util.Date.class };
+				case SOURCE, REASON -> new Class<?>[]{ String.class };
+				default -> null;
+			};
+		} else if (mode == ChangeMode.ADD || mode == ChangeMode.REMOVE) {
+			if (entryType == BanEntryType.EXPIRE_DATE)
+				return new Class<?>[]{ Timespan.class };
+		}
+		return null;
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Object banTarget = banTargetExpr.getSingle(event);
+		BanEntry<?> banEntry;
+
+		if (banTarget instanceof String ipTarget)
+			banEntry = getBanEntry(ipTarget);
+		else if (banTarget instanceof OfflinePlayer playerTarget)
+			banEntry = getBanEntry(playerTarget);
+		else
+			return;
+
+		if (banEntry == null)
+			return;
+
+		if (mode == ChangeMode.SET) {
+			switch (entryType) {
+				case SOURCE -> {
+					String newSource = (String) delta[0];
+					if (newSource == null)
+						return;
+					banEntry.setSource(newSource);
+				}
+				case EXPIRE_DATE -> {
+					ch.njol.skript.util.Date newDate = (ch.njol.skript.util.Date) delta[0];
+					if (newDate == null)
+						return;
+					banEntry.setExpiration(new java.util.Date(newDate.getTime()));
+				}
+				case REASON -> {
+					String newReason = (String) delta[0];
+					if (newReason == null)
+						return;
+					banEntry.setReason(newReason);
+				}
+			}
+		} else {
+			Timespan timespan = (Timespan) delta[0];
+			if (timespan == null)
+				return;
+
+			if (entryType == BanEntryType.EXPIRE_DATE) {
+				java.util.Date expiration = banEntry.getExpiration();
+				if (expiration == null)
+					return;
+				long newExpirationMillis;
+				if (mode == ChangeMode.ADD)
+					newExpirationMillis = expiration.getTime() + timespan.getAs(Timespan.TimePeriod.MILLISECOND);
```

Comment:
you may need to account for overflow here. you could see about using the Timespan operations registered with the Arithmetics class